Fix manual generation of config.json

rhass opened the pull request:
This fixes two issues with the generation of the docker/config.json. This makes the assumption that python is installed... which may or may not be true.

1. The generated JSON is now proper JSON. The previous output would loose the quotes on the key names and values.
2. Fixes a race condition when the plugin is called two times in a pipeline which causes the last credential to be the only credential in the file. The credentials now are appended to the hash each time.

rhass committed:
Fix manual generation of config.json

This fixes two issues with the generation of the docker/config.json.

1. The generated JSON is now proper JSON. The previous output would
   loose the quotes on the key names.
2. Fixes a race conditoin  when the plugin is called two times in a
   pipeline which causes the last credential to be the only credential
   in the file. The credentials now are appended to the hash each time.

rhass commented on hooks/generate-config.py:
```diff
@@ -0,0 +1,38 @@
+
+log.basicConfig(level=log.INFO, stream=sys.stderr)
+
+config_file_path = sys.argv[1] if len(sys.argv) > 1 else path.join(path.expanduser('~'), 'config.json')  
```
Oops... missing a path parameter.

rhass committed:
Add .docker to the default path.

rhass committed:
Downgrade log level from error to warn.

It is handled without a failure so error seems inappropriate here.

jonahbull commented:
@rhass I think you want feedback from `@elastic/release-eng`, they're the owners/maintainers of this plugin.

chloeruka commented:
> This makes the assumption that python is installed... which may or may not be true.

In this case I believe Python should always be installed on Buildkite agents as part of our standard tooling, @jarpy can you confirm?

rhass committed:
Move serialized data constructor out of file write.

This didn't really need to be in the file handler since it's just
updating a variable with the appropriate data... so I decided to move
this block.

alpar-t reviewed (commented):
I'm not sure about requiring python, not all CI workers will have it which limits the use of the plugin.  We could do it in a way where if python is available we can do better  maybe ?

alpar-t reviewed (commented):
Note that the part you are changing is really just a fall-back that we at some point even debated having. It will only be reached if you don't have any tools, like docker installed, so a better solution might be to remove this fallback and  require that one of the supported tools like skopeo or docker is installed in the worker.

rhass commented:
@alpar-t I fixed my issue by installing docker in the image I was using, however the two bugs in this code still remain without this fix or something similar. I don't think there is an easy way to solve the race condition issue without a non-standard external dependency since we would need a way to add k/v pairs to the `auths` hash key. To me, python is far more likely to be installed than `jq` which is why I opted for this solution.

All this being said, it may be better to just fail fast when docker isn't installed.

alpar-t commented:
I agree, would be better to fail than this fallback that doesn't work very well

jamietanna commented on hooks/generate-config.py:
```diff
@@ -0,0 +1,38 @@
+    }
+else:
+    config['auths'][environ['HOSTNAME']] = {
+        'auth': base64.b64encode(f"{environ['DOCKER_USERNAME']}:{environ['DOCKER_PASSWORD']}".encode('utf-8')).decode('utf-8')
```
Why the encode then decode?

rhass replied:
I can drop that as I don't think it's super necessary, but it's a crude way to sanitizing the inputs.

rhass commented:
I am going to close this as I think the correct thing to do is remove the fall-through case entirely.
